### PR TITLE
RFC: Support redef'ing of record field attributes

### PR DESCRIPTION
**Pull request by awelzel**
I'm not sure this is very useful outside of the &log attribute, but for
that it would allow users to remove &log from `Notice::Info$email_dest`
 without impacting any of the existing logic. Or, `&log` could be removed
from the base scripts and users can opt-in logging by redef'ing the field
with the `&log` attribute.

Relates to #2000.

### CODE REVIEW

**Comment by 0xxon** [comment]
Sounds great to me - the ability to add/remove attributes is something that has been missing for too long; the syntax looks fine to me.

**Comment by awelzel** [comment]
>  Sounds great to me - the ability to add/remove attributes is something that has been missing for too long; the syntax looks fine to me.

Thanks for looking! I had chatted a bit with @rsmmr and except for the `&log` attribute the others may be eye-brow raising questionable: Adding and more so removing `&default` or `&optional`? Changing the default value from a  `&default` would also be quite special.

I'll respin this with just supporting `&log` and erring for the others at this point.  There was mentioning that `&log` was always special in that it's not part of the language, but belongs more with the logging framework. Allowing more attributes later on is easier than removing it again.

The use case of adding `&log` to existing fields in `Info` records isn't as prevalent (most fields in Info are `&log` anyway), but there could be a new model of adding columns that aren't `&log` by default to Info record types. These can then be opted-in in via `redef += { &log }` rather than adding a `&log` into a base scripts for maybe a minority of users.

Possible other scenario might be 3rd-party packages where the author added tons of columns to a log, but one would like to only select a handful without patching the 3rd-party package.

Long story short: `&log` only for now :-)

**Comment by rsmmr** [comment]
I like the use cases that this enables (the `+=` even more than than the `-=`: as Arne wrote, that would allow us leaving new fields out of the logs by default if value is small, but still enable people to get them in there if they wanted). 

I somewhat torn on the mechanism because we're extending the language feature just for these cases. At the same time, I can't think of a better way to achieve this either, so I'm fine with it. On the plus side, we could extend this later to similar syntax for modifying attributes in other contexts than just record fields (e.g., modify expire timeout on tables).

So:

> Long story short: &log only for now :-) 

👍 from me

**Comment by J-Gras** [comment]
> I somewhat torn on the mechanism because we're extending the language here just for these cases. At the same time, I can't think of a better way to achieve this either, so I'm fine with it.

If it is just about logging, maybe [`Log::Filter`](https://docs.zeek.org/en/master/scripts/base/frameworks/logging/main.zeek.html#type-Log::Filter) are an option. They already support to exclude fields. Slightly adapting how `include` works or introducing something like `additional_fields` might work to support the discussed use cases.

**Comment by awelzel** [comment]
> If it is just about logging, maybe [`Log::Filter`](https://docs.zeek.org/en/master/scripts/base/frameworks/logging/main.zeek.html#type-Log::Filter) are an option. They already support to exclude fields. Slightly adapting how `include` works or introducing something like `additional_fields` might work to support the discussed use cases.

Yep. `$exclude` is in fact the workaround for #2000 on two independent Zeek distributions. It's a bit unfortunate, however, because filters can be created dynamically and each have individual `$exclude` / `$include` settings. Depending on situation (think user provided scripts creating filters) this needs to be considered individually for each filter instantiation.

Somewhat related, `ssl/files.zeek` supports redef's to "include" certificate subjects back into the ssl.log. It's implemented by adding those fields to the `$exclude` set of the default filter:

https://github.com/zeek/zeek/blob/master/scripts/base/protocols/ssl/files.zeek#L108-L123

There is a limitation here that those options are only in effect for the default filter. If you create filters manually, the options are not meaningful unless the same logic is implemented again (or the default filter is cloned). Adding and removing `&log`  for the fields in question gives an alternative approach. E.g. could remove &log from the base scripts and if someone really wants them, ask them to redef them with `&log` again. Whether that's better, not sure, but it's an alternative.

There's also the topic of log schema generation #1742 - if implemented by traversing records looking for `&log` attributes, the downstream filtering on the filters wouldn't be visible and the produced schema may have fields that aren't actually part of the log.

Thanks for your thoughts. But think I'll go ahead.

**Review by rsmmr — APPROVED** [review]
